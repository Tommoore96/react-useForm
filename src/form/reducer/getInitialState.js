import { fromJS, Map, List } from 'immutable'
import { fieldsKey, listenersKey } from './fieldReducer'
import { getFieldState } from './getFieldState'

/**
 * @param fields {Array<import("..").IFormFieldArgs>}
 */
export const getInitialState = (fields, initialValues = Map(), options = {}, formTools) => {
  const fieldMap = fields.reduce((acc, field) => {
    const result = acc.set(field.name, getFieldState(field, initialValues, options))

    return result
  }, Map())

  const listeners = getListeners(options.listeners)
  const withFormChange = options.onFormChange ? fromJS({ onFormChange: options.onFormChange }) : Map()
  return fromJS({ [fieldsKey]: fieldMap })
    .merge(listeners)
    .merge(withFormChange)
    .set('formTools', formTools)
}

const getListeners = (listeners = {}) => {
  return Object.entries(listeners).reduce((acc, [fieldName, value]) => {
    return acc.updateIn([listenersKey, fieldName], List(), current => {
      return Array.isArray(value) ? current.concat(value) : current.push(value)
    })
  }, Map())
}
