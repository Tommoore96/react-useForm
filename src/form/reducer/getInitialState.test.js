import { fromJS } from 'immutable'
import { getInitialState } from './getInitialState'

jest.mock('react')
jest.useFakeTimers()

let fields, initialValues, formTools, options
beforeEach(() => {
  formTools = 'form-tools'
  options = { op: 'tion' }
  fields = [
    { name: 'first', label: 'First' },
    { name: 'second', label: 'Second', type: 'boolean' },
  ]
  initialValues = fromJS({
    first: 'firstValue',
    second: 'secondValue',
  })

})

test('prepares initialState by field', () => {
  expect(getInitialState(fields, initialValues, undefined, formTools)).toMatchSnapshot()
})

test('throws error if name not provided', () => {
  fields.push({ label: 'Third' })
  expect(() => getInitialState(fields, initialValues, options, formTools)).toThrowErrorMatchingSnapshot()
})

test('adds helperText from field config', () => {
  fields[0].helperText = 'i am helper'
  expect(getInitialState(fields, initialValues, options, formTools)).toMatchSnapshot()
})

test('sets optional on field', () => {
  fields[0].optional = true
  expect(getInitialState(fields, initialValues, options, formTools)).toMatchSnapshot()
})

test('sets optional on field using custom formatter', () => {
  const optionalLabelFormatter = jest.fn()
  optionalLabelFormatter.mockReturnValue('label formatted')
  fields[0].optional = true
  expect(getInitialState(fields, initialValues, { optionalLabelFormatter }, formTools)).toMatchSnapshot()

  expect(optionalLabelFormatter.mock.calls[0]).toMatchSnapshot()
})

test('reads initialValues that are nested', () => {
  fields = [
    { name: 'nested.first', label: 'First' },
  ]
  initialValues = fromJS({
    nested: {
      first: 'nestedValue',
    },
  })
  expect(getInitialState(fields, initialValues, options, formTools)).toMatchSnapshot()
})

describe('initialValues', () => {
  test('field.value should used if set', () => {
    fields = [
      { name: 'first', type: 'text', value: 'i am field value' },
    ]
    expect(getInitialState(fields, undefined, undefined, formTools)).toMatchSnapshot()
  })

  test('text', () => {
    fields = [
      { name: 'first', type: 'text' },
    ]
    expect(getInitialState(fields, undefined, undefined, formTools)).toMatchSnapshot()
  })

  test('select', () => {
    fields = [
      { name: 'first', type: 'select' },
    ]
    expect(getInitialState(fields, undefined, undefined, formTools)).toMatchSnapshot()
  })

  test('number', () => {
    fields = [
      { name: 'first', type: 'number' },
    ]
    expect(getInitialState(fields, undefined, undefined, formTools)).toMatchSnapshot()
  })

  test('boolean', () => {
    fields = [
      { name: 'first', type: 'boolean' },
    ]
    expect(getInitialState(fields, undefined, undefined, formTools)).toMatchSnapshot()
  })

  test('list', () => {
    fields = [
      { name: 'first', type: 'list' },
    ]
    expect(getInitialState(fields, undefined, undefined, formTools)).toMatchSnapshot()
  })
})

test('adds field listeners', () => {
  const firstListener = jest.fn()
  const secondListenerA = jest.fn()
  const secondListenerB = jest.fn()
  const options = {
    listeners: {
      first: firstListener,
      second: [secondListenerA, secondListenerB],
    },
  }

  expect(getInitialState(fields, initialValues, options, formTools)).toMatchSnapshot()
})

test('adds onFormChange when present', () => {
  options.onFormChange = 'on-form-change'
  expect(getInitialState(fields, initialValues, options, formTools)).toMatchSnapshot()
})
