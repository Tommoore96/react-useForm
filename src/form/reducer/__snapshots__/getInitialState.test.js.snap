// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds field listeners 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "optional": false,
        "label": "First",
        "field": Object {
          "label": "First",
          "name": "first",
        },
        "value": "firstValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
        "value": "secondValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
  "listeners": Immutable.Map {
    "first": Immutable.List [
      [MockFunction],
    ],
    "second": Immutable.List [
      [MockFunction],
      [MockFunction],
    ],
  },
  "formTools": "form-tools",
}
`;

exports[`adds helperText from field config 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "optional": false,
        "label": "First",
        "field": Object {
          "helperText": "i am helper",
          "label": "First",
          "name": "first",
        },
        "value": "firstValue",
      },
      "current": Immutable.Map {
        "helperText": "i am helper",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
        "value": "secondValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
  "formTools": "form-tools",
}
`;

exports[`adds onFormChange when present 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "optional": false,
        "label": "First",
        "field": Object {
          "label": "First",
          "name": "first",
        },
        "value": "firstValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
        "value": "secondValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
  "onFormChange": "on-form-change",
  "formTools": "form-tools",
}
`;

exports[`initialValues boolean 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "boolean",
        },
        "value": false,
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": false,
      },
    },
  },
  "formTools": "form-tools",
}
`;

exports[`initialValues field.value should used if set 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "text",
          "value": "i am field value",
        },
        "value": "i am field value",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "i am field value",
      },
    },
  },
  "formTools": "form-tools",
}
`;

exports[`initialValues list 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "list",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "list",
        },
        "options": Object {},
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
      },
      "items": Immutable.List [],
    },
  },
  "formTools": "form-tools",
}
`;

exports[`initialValues number 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "number",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "number",
        },
        "value": 0,
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": 0,
      },
    },
  },
  "formTools": "form-tools",
}
`;

exports[`initialValues select 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "select",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "select",
        },
        "value": "",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "",
      },
    },
  },
  "formTools": "form-tools",
}
`;

exports[`initialValues text 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "optional": false,
        "label": undefined,
        "field": Object {
          "name": "first",
          "type": "text",
        },
        "value": "",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "",
      },
    },
  },
  "formTools": "form-tools",
}
`;

exports[`prepares initialState by field 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "optional": false,
        "label": "First",
        "field": Object {
          "label": "First",
          "name": "first",
        },
        "value": "firstValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
        "value": "secondValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
  "formTools": "form-tools",
}
`;

exports[`reads initialValues that are nested 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "nested.first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "optional": false,
        "label": "First",
        "field": Object {
          "label": "First",
          "name": "nested.first",
        },
        "value": "nestedValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "nestedValue",
      },
    },
  },
  "formTools": "form-tools",
}
`;

exports[`sets optional on field 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "optional": true,
        "label": "First (optional)",
        "field": Object {
          "label": "First",
          "name": "first",
          "optional": true,
        },
        "value": "firstValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
        "value": "secondValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
  "formTools": "form-tools",
}
`;

exports[`sets optional on field using custom formatter 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "first": Immutable.Map {
      "initial": Immutable.Map {
        "type": "text",
        "optional": true,
        "label": "label formatted",
        "field": Object {
          "label": "First",
          "name": "first",
          "optional": true,
        },
        "value": "firstValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "firstValue",
      },
    },
    "second": Immutable.Map {
      "initial": Immutable.Map {
        "type": "boolean",
        "optional": false,
        "label": "Second",
        "field": Object {
          "label": "Second",
          "name": "second",
          "type": "boolean",
        },
        "value": "secondValue",
      },
      "current": Immutable.Map {
        "helperText": "",
        "error": false,
        "pristine": true,
        "touched": false,
        "value": "secondValue",
      },
    },
  },
  "formTools": "form-tools",
}
`;

exports[`sets optional on field using custom formatter 2`] = `
Array [
  "First",
  "first",
]
`;

exports[`throws error if name not provided 1`] = `"name is required on field"`;
